### ISSUE TITLE
3D prediction memory leak?

### ISSUE THREAD

**Issue by Optimox** [stardist/stardist#130]
Hello again,

I noticed some erratic behavior during prediction (not evaluation).

I have a 3D stardist model and I individually loop over some 3D stacks of shapes around : (355, 500, 1200)

I simply call this in my for loop then save results on the disk:
```
labels, details = model.predict_instances(stack,
                                              prob_thresh=PROB_THRESH,
                                              nms_thresh=NMS_THRESH,
                                              n_tiles = (4,5,10)
                                                 )
```

The GPU inference part is running smoothly, but after this (I guess during the NMS) I see my RAM peaking to > 40Go (I mean that this process takes more than 40Go of RAM while I have 64). This sometimes kills my process, the same 3D stack might once kill the process while it will do ok at next try (I don't have anything heavy running on the side).

It sometimes runs ok for 60 consecutive stacks, then stop, sometime it's after only 5 stacks.

I'm wondering if I'm doing something wrong? Is it supposed to take that much RAM to make the inference post processing?

Should I add more tiles? Is the post processing tile dependent?

I'm running with tensorflow 2.3.1 and startdist 0.6.2.

Thanks for your help.

Edit:
After a fresh restart and nothing else running, I see between 8Go and 20Go of RAM used during inference on GPU ( 5G of GPU MEM is used) and then just after inference on GPU is done there is a peak to 44Go of RAM for one second then back to 25Go. One stack takes about 400Mb on disk (int8) so I do understand that float 32 and a few copies can bring this to 5-10 Go, but 20-40 seems too heavy. I have 24 cpu cores and they all seem to be used at 100% by the post-processing, could that be the reason for memory usage? Can I limit the number of thread to be used for post-processing?

**Comment by uschmidt83** [comment]
Hi @Optimox,

> The GPU inference part is running smoothly, but after this (I guess during the NMS) I see my RAM peaking to > 40Go (I mean that this process takes more than 40Go of RAM while I have 64). This sometimes kills my process...

> I'm wondering if I'm doing something wrong? Is it supposed to take that much RAM to make the inference post processing?

The NMS postprocessing can use a lot of RAM. How much exactly will also depend on the content of your images and the chosen prob and nms thresholds. We will be alleviating this issue with the next release of StarDist, please see this [pull request](https://github.com/stardist/stardist/pull/124) for details. In the meantime, if you're feeling adventurous you can get the benefits today by installing the StarDist `multi` branch. Or you can use `model.predict_instances_big` (available in the current version of StarDist), which will also do the postprocessing in overlapping blocks. Please see this [example notebook](https://nbviewer.jupyter.org/github/stardist/stardist/blob/master/examples/other2D/predict_big_data.ipynb) for details, and don't worry it also works in 3D.

> Should I add more tiles? Is the post processing tile dependent?

No, only if you use `model.predict_instances_big`, which introduces additional parameters for that purpose.

> I have 24 cpu cores and they all seem to be used at 100% by the post-processing, could that be the reason for memory usage? Can I limit the number of thread to be used for post-processing?

No, limiting the number of threads won't lower memory usage. However, you can (for instance) set the environment variable `OMP_NUM_THREADS=8` before running StarDIst to limit the number of threads to 8.

**Comment by Optimox** [comment]
Thank you for your answer, great to see that this will be improved soon.

I've got a question about predict instances big, how does this function deal with borders of tiles? Do you overlap tiles and then average the results on overlapping regions? Do you expect the model to be slightly worse on tiles borders? (for example you could have more overlapping cells that the NMS threshold should allow on borders no?)

**Comment by uschmidt83** [comment]
Just so that we're talking about the same things, the StarDist API refers to tiles when it comes to breaking up images for neural network prediction, and `predict_instances_big` denotes as blocks those regions that are independently processed, including NMS. (Each block can be broken up into several tiles for the neural net prediction.)

Neighboring blocks are overlapped in such a way that each segmented object is fully contained in at least one of the blocks. Each block knows which objects it is responsible for, such that the result of `predict_instances_big` and the "normal" `predict_instances` should be identical (up to permutation of the label ids). It's all a bit complicated under the hood, and I meant to make a little diagram to explain it visually. Maybe I'll get around to make this some day.
